src/test/java/com/snowflake/kafka/connector/internal/streaming/StreamingClientProviderConcurrencyTest.java

```diff
@@ -0,0 +1,303 @@
+import com.snowflake.kafka.connector.Utils;
+import com.snowflake.kafka.connector.internal.TestUtils;
+import java.util.*;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import net.snowflake.ingest.streaming.SnowflakeStreamingIngestClient;
+import org.junit.*;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.mockito.Mockito;
+
+@RunWith(Parameterized.class)
+public class StreamingClientProviderConcurrencyTest {
+  private Map<String, String> clientConfig;
+
+  private StreamingClientProvider streamingClientProvider;
+  private FakeStreamingClientHandler streamingClientHandler;
+  private boolean enableClientOptimization;
+
+  private List<Future<SnowflakeStreamingIngestClient>> getClientFuturesTeardown;
+  private List<Future> closeClientFuturesTeardown;
+
+  @Parameterized.Parameters(name = "enableClientOptimization: {0}")
+  public static Collection<Object[]> input() {
+    return Arrays.asList(new Object[][] {{true}, {false}});
+  }
+
+  public StreamingClientProviderConcurrencyTest(boolean enableClientOptimization) {
+    this.enableClientOptimization = enableClientOptimization;
+  }
+
+  @Before
+  public void setup() {
+    this.clientConfig = TestUtils.getConfForStreaming();
+    this.clientConfig.put(
+        SnowflakeSinkConnectorConfig.ENABLE_STREAMING_CLIENT_OPTIMIZATION_CONFIG,
+        this.enableClientOptimization + "");
+
+    this.streamingClientHandler = new FakeStreamingClientHandler();
+    this.streamingClientProvider =
+        StreamingClientProvider.getStreamingClientProviderForTests(
+            this.streamingClientHandler,
+            StreamingClientProvider.buildLoadingCache(this.streamingClientHandler));
+
+    this.getClientFuturesTeardown = new ArrayList<>();
+    this.closeClientFuturesTeardown = new ArrayList<>();
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    // clean up all the threads
+    try {
+      for (Future<SnowflakeStreamingIngestClient> getClientFuture : this.getClientFuturesTeardown) {
+        getClientFuture.get().close();
+      }
+      for (Future closeClientFuture : this.closeClientFuturesTeardown) {
+        closeClientFuture.get();
+      }
+    } catch (Exception ex) {
+      throw ex;
+    }
+  }
+
+  @Ignore
+  // SNOW-840882 flakey test
+  public void testMultipleGetAndClose() throws Exception {
+    // setup configs for 3 clients
+    Map<String, String> clientConfig1 = new HashMap<>(this.clientConfig);
+    Map<String, String> clientConfig2 = new HashMap<>(this.clientConfig);
+    Map<String, String> clientConfig3 = new HashMap<>(this.clientConfig);
+
+    clientConfig1.put(Utils.NAME, "client1");
+    clientConfig2.put(Utils.NAME, "client2");
+    clientConfig3.put(Utils.NAME, "client3");
+
+    int createClientCount = 0;
+
+    // task1: get client x3, close client, get client, close client
```

Comment:
Maybe creating a separate interface exposing the mentioned methods: `get client`, `close client`... would make tests more straightforward and shorter. Is it possible?